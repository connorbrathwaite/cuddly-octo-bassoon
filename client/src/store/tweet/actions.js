export const REQUEST_ADD_TWEET = 'REQUEST_ADD_TWEET'

export const SUCCESS_ADD_TWEET = 'SUCCESS_ADD_TWEET'

export const REQUEST_TWEETS = 'REQUEST_TWEETS'

export const SUCCESS_TWEETS = 'SUCCESS_TWEETS'

export const FAILURE_TWEETS = 'FAILURE_TWEETS'

export const requestAdd = (candidate, tweet) => ({
  type: REQUEST_ADD_TWEET,
  payload: {
    candidate,
    tweet
  }
})

export const add = (candidate, tweet) => ({
  type: SUCCESS_ADD_TWEET,
  payload: {
    candidate,
    tweet
  }
})

export const request = candidate => ({
  type: REQUEST_TWEETS,
  payload: {
    candidate
  }
})

export const success = (candidate, tweets) => ({
  type: SUCCESS_TWEETS,
  payload: {
    candidate,
    tweets
  }
})

export const failure = (candidate, error) => ({
  type: FAILURE_TWEETS,
  payload: {
    candidate,
    error
  }
})
