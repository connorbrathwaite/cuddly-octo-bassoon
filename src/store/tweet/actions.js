const REQUEST_TWEETS = 'REQUEST_TWEETS'
const SUCCESS_TWEETS = 'SUCCESS_TWEETS'
const FAILURE_TWEETS = 'FAILURE_TWEETS'

export const request = candidate => ({
  type: REQUEST_TWEETS,
  payload: {
    candidate
  }
})
export const success = (candidate, response) => ({
  type: SUCCESS_TWEETS,
  payload: {
    candidate,
    tweets: response.tweets
  }
})
export const failure = (candidate, error) => ({
  type: FAILURE_TWEETS,
  payload: {
    candidate,
    error
  }
})
