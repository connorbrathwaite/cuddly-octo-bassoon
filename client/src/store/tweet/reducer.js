import {
  cond,
  path,
  equals,
  always,
  evolve,
  union,
  T,
  F
} from 'ramda'
import * as actions from './actions'

const initialState = {
  isLoading: F(),
  hasError: F(),
  tweets: []
}

const reducer = (state = initialState, action) =>
  cond([
    [
      equals(actions.REQUEST_TWEETS),
      () =>
        evolve({
          isLoading: T,
          hasError: F
        })(state)
    ],
    [
      equals(actions.SUCCESS_TWEETS),
      () =>
        evolve({
          isLoading: F,
          hasError: F,
          tweets: union(action.payload.tweets)
        })(state)
    ],
    [
      equals(actions.FAILURE_TWEETS),
      () =>
        evolve({
          isLoading: F,
          hasError: T
        })(state)
    ],
    [T, always(state)]
  ])(action.type)

export default reducer
