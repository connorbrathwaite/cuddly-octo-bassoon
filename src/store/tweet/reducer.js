import {
  cond,
  equals,
  always,
  evolve,
  prepend,
  T,
  F
} from 'ramda'
import * as actions from './actions'

const initialState = {
  loading: false,
  error: false,
  tweets: []
}

const reducer = (state = initialState, action) =>
  cond([
    [equals(actions.request), always({loading: T()})],
    [
      equals(actions.success),
      () =>
        evolve({
          loading: F(),
          tweets: prepend(action.payload.tweets)
        })(state)
    ],
    [T, always(state)]
  ])(action.type)

export default reducer
