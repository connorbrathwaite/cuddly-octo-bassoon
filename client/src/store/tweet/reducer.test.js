import reducer, {initalState} from './reducer'
import {add, request, success, failure} from './actions'

const mockTweet = {
  id: 991675463184539600,
  author: "JanLandy",
  candidate: "trump",
  createdAt: "Wed May 02 13:47:09 +0000 2018",
  text:"@renato_mariotti I think we have to understand that Trump is guilty, and because he’s guilty he’s going to do every… https://t.co/3eOqwlBew0"
}

describe('tweet reducer', () => {
  test('recieve tweets', () => {
    const state = reducer(initalState, success('trump', [mockTweet]))
    expect(state.tweets).toMatchSnapshot()
    expect(state.isLoading).toBeFalsy()
    expect(state.hasError).toBeFalsy()
  })
})
