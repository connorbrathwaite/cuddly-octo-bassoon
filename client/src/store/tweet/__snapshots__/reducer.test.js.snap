// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tweet reducer recieve tweets 1`] = `
Array [
  Object {
    "author": "JanLandy",
    "candidate": "trump",
    "createdAt": "Wed May 02 13:47:09 +0000 2018",
    "id": 991675463184539600,
    "text": "@renato_mariotti I think we have to understand that Trump is guilty, and because he’s guilty he’s going to do every… https://t.co/3eOqwlBew0",
  },
]
`;
